### DISCUSSION TITLE
Fragments for CommonMark headings

### DISCUSSION THREAD

**Discussion by mattias-p** [community/community#21546] [New to GitHub]
<p>When GitHub renders CommonMark documents as HTML, the headings are augmented with fragment links back to the heading itself. I’ve looked for published documentation describing the algorithm that generates these links, but I couldn’t find anything authoritative. Is there such a document? If not, could it be published?</p>

**Comment by shawnajean** [comment, marked as answer]
<p>Hi <a class="mention" href="/mattias-p">@mattias-p</a>! :wave:t2:  These anchor links for markdown headings are actually added outside of the CommonMark renderer itself. We use the github/markup repository for rendering markup languages to HTML:</p>
<p><a href="https://github.com/github/markup#github-markup" rel="nofollow noopener">https://github.com/github/markup#github-markup</a></p>
<p>And then pass that content through various other filters in the html-pipeline (mentioned in Step 3 above). The ‘named anchors’ link there is what you’re looking for. Here it is for ease:</p>
<p><a href="https://github.com/jch/html-pipeline/blob/master/lib/html/pipeline/toc_filter.rb" rel="nofollow noopener">https://github.com/jch/html-pipeline/blob/master/lib/html/pipeline/toc_filter.rb</a></p>
<p>We don’t actually render a table of contents for markdown files, but we do use this to render the anchor links.</p>

**Comment by RyanZim** [comment]
<p>If you’re in the Node.js world, here’s a module that generates them in a way that’s compliant with GitHub: <a href="https://github.com/thlorenz/anchor-markdown-header" rel="nofollow noopener">https://github.com/thlorenz/anchor-markdown-header</a></p>

**Comment by mattias-p** [comment]
<p>Thank you for your answers! They were very helpful.</p>

**Comment by srl295** [comment]
years later, https://github.com/gjtorikian/html-pipeline/tree/main?tab=readme-ov-file#html-pipeline says that html-pipeline is no longer used by GitHub
